Lib/fontmake/instantiator.py

```diff
@@ -0,0 +1,579 @@
+                    raise InstantiatorError(
+                        f"Failed to generate instance of glyph '{glyph_name}'."
+                    ) from e
+
+                # ...except if the glyph is in public.skipExportGlyphs and would
+                # therefore be removed from the compiled font anyway. There's not much
+                # we can do except leave it empty in the instance and tell the user.
+                logger.warning(
+                    "Failed to generate instance of glyph '%s', which is marked as "
+                    "non-exportable. Glyph will be left empty. Failure reason: %s",
+                    glyph_name,
+                    e,
+                )
+
+            glyph.unicodes = [uv for uv in self.glyph_name_to_unicodes[glyph_name]]
+
+        # Process rules
+        glyph_names_list = self.glyph_mutators.keys()
+        glyph_names_list_renamed = designspaceLib.processRules(
+            self.designspace_rules, location, glyph_names_list
+        )
+        for name_old, name_new in zip(glyph_names_list, glyph_names_list_renamed):
+            if name_old != name_new:
+                swap_glyph_names(font, name_old, name_new)
+
+        return font
+
+    def _generate_instance_info(
+        self,
+        instance: designspaceLib.InstanceDescriptor,
+        location_normalized: Location,
+        location: Location,
+        font: ufoLib2.Font,
+    ) -> None:
+        """Generate fontinfo related attributes.
+
+        Separate, as fontinfo treatment is more extensive than the rest.
+        """
+        info_instance = self.info_mutator.instance_at(location_normalized)
+        if self.round_geometry:
+            info_instance = info_instance.round()
+        info_instance.extractInfo(font.info)
+
+        # Copy non-interpolating metadata from the default font.
+        for attribute in ufoLib.fontInfoAttributesVersion3:
+            if hasattr(info_instance, attribute):
+                continue  # Skip interpolated info attributes.
+            if hasattr(self.copy_info, attribute):
+                setattr(
+                    font.info,
+                    attribute,
+                    copy.deepcopy(getattr(self.copy_info, attribute)),
+                )
+        for key, glyph_names in self.copy_groups.items():
+            font.groups[key] = [name for name in glyph_names]
+        font.features.text = self.copy_feature_text
+        # 1. Copy the default lib to the instance.
+        font.lib = typing.cast(dict, copy.deepcopy(self.copy_lib))
+        # 2. Copy the Designspace instance lib to the instance, overwriting the default.
+        for key, value in instance.lib.items():
+            font.lib[key] = copy.deepcopy(value)
+        # 3. Copy the Designspace public.skipExportGlyphs lib key to the instance lib
+        #    to make sure the Designspace version wins.
+        font.lib["public.skipExportGlyphs"] = [name for name in self.skip_export_glyphs]
```

Comment:
we should let an instance-specific "public.skipExportGlyphs" list from Designspace.instance.lib take precedence over a global one from designspace.lib.

Also this behavior whereby designspace.instances.lib elements are copied to the generated UFO instance should be documented in the Designspace spec (I don't think it is, currently, beyond a generic "custom data for this instance").

Reply:
1. Done
2. https://github.com/fonttools/fonttools/pull/1683